roles/postgresql/tasks/upgrade.yml

```diff
@@ -0,0 +1,111 @@
+---
+# copyright Utrecht University
+
+- name: Check for presence of PostgreSQL 9
+  ansible.builtin.package_facts:
+    manager: auto
+
+
+- name: Stop PostgreSQL 9 service
+  ansible.builtin.service:
+    name: postgresql
+    enabled: false
+    state: stopped
+  when: "'postgresql-server' in ansible_facts.packages"
+
+
+- name: Create PostgreSQL 9 shim directory
+  ansible.builtin.file:
+    path: /var/lib/pgsql/shim
+    owner: postgres
+    group: postgres
+    state: directory
+  when: "'postgresql-server' in ansible_facts.packages"
+
+
+- name: Install PostgreSQL 9 shims
+  ansible.builtin.file:
+    src: "/usr/bin/{{ item }}"
+    dest: "/var/lib/pgsql/shim/{{ item }}"
+    owner: postgres
+    group: postgres
+    state: link
+  with_items:
+    - clusterdb
+    - createdb
+    - createuser
+    - dropdb
+    - dropuser
+    - initdb
+    - oid2name
+    - pg_archivecleanup
+    - pg_basebackup
+    - pgbench
+    - pg_config
+    - pg_controldata
+    - pg_dump
+    - pg_dumpall
+    - pg_receivexlog
+    - pg_resetxlog
+    - pg_restore
+    - pg_standby
+    - pg_test_fsync
+    - pg_test_timing
+    - postgres
+    - postgresql-check-db-dir
+    - postgresql-setup
+    - postmaster
+    - psql
+    - reindexdb
+    - vacuumdb
+    - vacuumlo
+  when: "'postgresql-server' in ansible_facts.packages"
+
+
+- name: Install PostgreSQL 9 shim script
+  ansible.builtin.copy:
+    src: "pg_ctl"
+    dest: /var/lib/pgsql/shim/pg_ctl
+    owner: postgres
+    group: postgres
+    mode: 0755
+  when: "'postgresql-server' in ansible_facts.packages"
+
+
+- name: Last step before the upgrade
```

Comment:
Moving this failsafe up to the top of the upgrade tasks would prevent it from leaving postgres in a stopped state. Additionally, I think it would be safer to always fail if prevent_upgrade is enabled (either "when: prevent_upgrade" or "when prevent_upgrade or ...")